### ISSUE TITLE
Having custom validation over scalars

### ISSUE THREAD

**Issue by mxamin** [graphql-python/graphene#1070]
I need to pass some validation check when defining `Arguments` for Mutation. I know I can check it inside the `mutate` function but I need this check before that (similar to what we have in `marshmallow.validate` module: `Range` and `Length`). 

I just want to know if there is any future plan to implement it inside the graphene or not.

**Comment by KaySackey** [comment]
You can create a custom scalar that does validation. It’s what I do anyways, when accepting input strings to validate that user data doesn’t contain HTML.

**Comment by mxamin** [comment]
But the scalars won't accept inputs, suppose I want to define a new scalar which accepts strings with specific lengths, in that case, I want to pass the length value to my custom scalar but as far as I found out, scalars won't accept input parameters and you have to hardcode the length.
Correct me if I'm wrong :)

**Comment by KaySackey** [comment]
Not sure what you are doing in particular, but Scalars can accept kwargs. You just have to access them via `self.kwargs`

**Comment by chpmrc** [comment]
Sure, you can pass kwargs but where exactly would the validation logic run and be called? `parse_value`, `serialize`? Also there's no access to the supertype so what if I wanted to show a hierarchy/path to the wrong value (e.g. `items[0].email is invalid`)?

**Comment by jkimbo** [comment]
@chpmrc do you need to validate before the `mutate` function? (just trying to understand the use case better)

**Comment by chpmrc** [comment]
Yes, similarly to how Django validates input in the forms, before calling the various `create`, `update`, etc. I threw together a decorator for pre mutation validation for graphene, I'll see if I can open a PR once it's polished.

**Comment by jkimbo** [comment]
I was going to suggest using a decorator! I think that is the easiest approach to add validation.

I'd be interested in seeing a PR if you get it working.

**Comment by chpmrc** [comment]
@jkimbo so I came up with this PoC and I was hoping you could take a look to maybe integrate it in Graphene (after some refinements): https://github.com/chpmrc/graphene-validator . Let me know what you think! (I put the repo together quickly but we're already using it in production)

**Comment by chpmrc** [comment]
@jkimbo did you have a chance to look at the repo I linked above? Cheers!

**Comment by vancouverwill** [comment]
> Sure, you can pass kwargs but where exactly would the validation logic run and be called? `parse_value`, `serialize`? Also there's no access to the supertype so what if I wanted to show a hierarchy/path to the wrong value (e.g. `items[0].email is invalid`)?

@chpmrc I got the validation we needed working by using a custom Scalar and adding the validation on `parse_value()` as that does get called by the `Graphene` library before input is passed to either query or mutations. Adding an example here as the docs are super sparse and might of help to you or someone else in the future

```
class Email(String):

    @staticmethod
    def parse_value(value: Text) -> Text:
        if is_valid_email(value):
            return value
        raise GraphQLError(
                f"Invalid Graph ID {value}, should be hexadecimal format 0x1f1f"
            )

def is_valid_email(email):
    # do some regex or other validation
```

**Comment by chpmrc** [comment]
@vancouverwill thanks for that example! Does it return the `path` to the invalid field in the error? E.g. if you send an object `{user: {name, email}}` will `path` be `[user, email]` in case of invalid email?

**Comment by hypnotic-frog** [comment]
> parse_value

This doesn't work for me. I used exactly the same example as above and the parse_value function never gets called. How is this possible? Anyone has found a reliable way to add custom validator to a custom scalar type?

Thanks!!

**Comment by meysam81** [comment]
@hypnotic-frog 

Turn `parse_value` to `parse_literal` and it'll work. Tested and got the result.

**Comment by aryadovoy** [comment]
Is it possible to pass `info` in `parse_value()`?

And how can I raise multiple errors after validations from all inputs?
